Use of Atomic64 in minijinja-contrib prevents minijinja-cli building on some 32-bit platforms

## Description

Related bug: https://github.com/mitsuhiko/minijinja/issues/207

Building `minijinja-cli` on a 32-bit platform fails:

```
    Checking minijinja-contrib v2.8.0 (/home/wmoore/Source/github.com/mitsuhiko/minijinja/minijinja-contrib)
error[E0432]: unresolved import `std::sync::atomic::AtomicU64`
 --> minijinja-contrib/src/rand.rs:3:25
  |
3 | use std::sync::atomic::{AtomicU64, Ordering};
  |                         ^^^^^^^^^
  |                         |
  |                         no `AtomicU64` in `sync::atomic`
  |                         help: a similar name exists in the module: `AtomicU32`
```

## Reproduction steps

1. `rustup target add armv5te-unknown-linux-gnueabi`
2. `cargo check -p minijinja-cli --target armv5te-unknown-linux-gnueabi`

Additional helpful information:

- Version of minijinja: a258405
- Version of rustc: rustc 1.85.0 (4d91de4e4 2025-02-17)
- Operating system and version: Arch Linux x86_64

This was originally noticed in the `minijinja-cli` package for Chimera Linux, which has 32-bit PPC and ARM targets:

https://github.com/chimera-linux/cports/compare/a846a17534f2...454355c25169

## What did you expect

`minijinja-cli` to compile.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -15,6 +15,8 @@ All notable changes to MiniJinja are documented here.
 - Added `mj_value_new_bytes` to the C-ABI.  #749
 - Added `mj_value_as_bytes` to the C-ABI to borrow from strings or
   byte values.  #750
+- Fix a compilation issue on 32bit systems when `AtomicU64` is
+  not available in minijinja-contrib.  #755
 
 ## 2.8.0
 

--- a/minijinja-contrib/src/rand.rs
+++ b/minijinja-contrib/src/rand.rs
@@ -1,6 +1,5 @@
 use std::collections::hash_map::RandomState;
 use std::hash::{BuildHasher, Hasher};
-use std::sync::atomic::{AtomicU64, Ordering};
 use std::sync::Arc;
 
 use minijinja::value::{Object, ObjectRepr};
@@ -8,7 +7,7 @@ use minijinja::State;
 
 #[derive(Debug)]
 pub struct XorShiftRng {
-    seed: AtomicU64,
+    seed: seed_impl::Seed,
 }
 
 impl Object for XorShiftRng {
@@ -30,18 +29,18 @@ impl XorShiftRng {
 
     pub fn new(seed: Option<u64>) -> XorShiftRng {
         XorShiftRng {
-            seed: AtomicU64::from(
+            seed: seed_impl::Seed::new(
                 seed.unwrap_or_else(|| RandomState::new().build_hasher().finish()),
             ),
         }
     }
 
     pub fn next(&self) -> u64 {
-        let mut rv = self.seed.load(Ordering::Relaxed);
+        let mut rv = seed_impl::load(&self.seed);
         rv ^= rv << 13;
         rv ^= rv >> 7;
         rv ^= rv << 17;
-        self.seed.store(rv, Ordering::Relaxed);
+        seed_impl::store(&self.seed, rv);
         rv
     }
 
@@ -55,5 +54,31 @@ impl XorShiftRng {
 
     pub fn random_range(&self, lower: i64, upper: i64) -> i64 {
         (self.random() * (upper - lower) as f64) as i64 + lower
+    }
+}
+
+#[cfg(target_has_atomic = "64")]
+mod seed_impl {
+    pub type Seed = std::sync::atomic::AtomicU64;
+
+    pub fn load(seed: &Seed) -> u64 {
+        seed.load(std::sync::atomic::Ordering::Relaxed)
+    }
+
+    pub fn store(seed: &Seed, v: u64) {
+        seed.store(v, std::sync::atomic::Ordering::Relaxed);
+    }
+}
+
+#[cfg(not(target_has_atomic = "64"))]
+mod seed_impl {
+    pub type Seed = std::sync::Mutex<u64>;
+
+    pub fn load(seed: &Seed) -> u64 {
+        *seed.lock().unwrap()
+    }
+
+    pub fn store(seed: &Seed, v: u64) {
+        *seed.lock().unwrap() = v;
     }
 }

